aten/src/ATen/Context.h

```diff
@@ -204,7 +208,15 @@ static inline bool hasHIP() {
   return globalContext().hasHIP();
 }
 
+static inline bool hasXLA() {
+  return globalContext().hasXLA();
+}
+
 static inline size_t getNumGPUs() {
+  auto xla_hooks = detail::getXLAHooks();
+  if (xla_hooks != nullptr) {
+    return xla_hooks->num_devices();
```

Comment:
this means essentially I cannot have both XLA devices and GPU devices in the same pytorch instance (or at least, I won't have enough autograd threads).  We make this assumption for HIP because that's a limitation of how that specific part of PyTorch is written, but I'm not sure we really want to encode that elsewhere.

Is there anything preventing XLA and GPUs from coexisting?  Can we just add a new function getTotalNonCPUDevices or something that autograd can call?

Reply:
Good point, but unfortunately it gets tricky WRT autograd, and its device->queue indexing if we start support GPU and XLA.
It's even more complex since XLA itself, can lower for GPU :)

Reply:
well we'd need to index with actual devices, right?

> It's even more complex since XLA itself, can lower for GPU :) 

what does this mean exactly?

Reply:
Look at how the thread initializes.
Each thread sets up a device guard with specific device type.
That guard is then used in other places.
If we allow multiple devices PT GPU together with PT XLA, the autograd code will have to go under a pretty major change, as autograd lower lever code seems only to deal with device indices (and assume CUDA).

XLA internally can use GPU in theory, via the XLA paths (not PT).
From a user POV you could be using GPU, by using XLA, w/out noticing (same interface).
XLA also handle CPU, this is the way we do most of the development in our workstations.

Reply:
> If we allow multiple devices PT GPU together with PT XLA, the autograd code will have to go under a pretty major change, as autograd lower lever code seems only to deal with device indices (and assume CUDA).

Sure, that makes sense.  I can try to see how feasible making that kind of change is.

> XLA internally can use GPU in theory, via the XLA paths (not PT).
From a user POV you could be using GPU, by using XLA, w/out noticing (same interface).
XLA also handle CPU, this is the way we do most of the development in our workstations.

Thanks, that makes sense.  I guess the question is what is the right thing to do from autograd's perspective.  Is the assumption in this world that XLA controls all the (available) devices, so it would be bad perf wise or even invalid if the user is also directly using the GPUs via PyTorch?

Reply:
The right, long term, change for autograd would be to make it reason about Device everywhere. Like Adam was mentioning.
I took a quick peek at it, and mong the places where device guard info are used, and the places where only the device ordinal travels, seems pretty big change to me.